src/bin/scryer-prolog.rs

```diff
@@ -8,6 +9,20 @@ fn main() -> std::process::ExitCode {
     })
     .unwrap();
 
-    let mut wam = machine::Machine::new();
-    wam.run_top_level()
+    #[cfg(target_os = "wasi")]
+    let runtime = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+
+    #[cfg(not(target_os = "wasi"))]
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+
+    runtime.block_on(async move {
```

Comment:
I'm not sure what's the purpose of this if everything inside is sync (?)

Reply:
This block seems to be the issue of the errors reported by @mthom . Removing it makes all the test pass, but of course `http_listen/2` crashes. There are two options that I see:

- Go full mode async. This a more disruptive change I think. It's making the Machine async-aware, so system calls can be async, which could imply rewriting the `http_listen/2` parts (but I don't mind)
- Return to the old system of having a runtime attached to a Machine, which can be used to launch async tasks in sync contexts.

Reply:
I don't know enough about async or tokio to make a judgment as to which course is best. Can there not be simultaneous tokio runtimes? Or runtimes within runtimes?

Reply:
After tinkering more with this, there's a third option. Change the `tokio::sync::RwLock` to `std::sync::RwLock`. This is a very small change (see https://github.com/aarroyoc/scryer-prolog/tree/library-use-case) and everything seems to work again (both tests and `http_listen/2`). The ultimate reason seems to be that the machine executes now in an async environment (I still don't know why) and `RwLock::blocking_read/blocking_write` are expected to panic when called in an async environment since they're designed for sync code (https://docs.rs/tokio/latest/tokio/sync/struct.RwLock.html#method.blocking_read)

Reply:
Rust Futures need be polled in order to execute anything. That means you need at least one Future executor like `block_on` to keep polling the Futures to completion. If you would have this async block here in this main function, the app would return and never even start processing the async code.